ide/svg/src/org/netbeans/modules/svg/SVGViewerElement.java

```diff
@@ -0,0 +1,351 @@
+    }
+
+    @Override
+    public JComponent getVisualRepresentation() {
+        if (component == null) {
+            viewer = new JPanel();
+            component = viewer;
+        }
+
+        return component;
+    }
+
+    @Override
+    public JComponent getToolbarRepresentation() {
+        if (toolbar == null) {
+            toolbar = new SVGViewerToolbar().createToolbar(this);
+        }
+
+        return toolbar;
+    }
+
+    @Override
+    public Action[] getActions() {
+        return new Action[0];
+    }
+
+    @Override
+    public Lookup getLookup() {
+        return dataObject.getLookup();
+    }
+
+    @Override
+    public void componentOpened() {
+        dataObject.getPrimaryFile().addFileChangeListener(fcl);
+
+        updateView();
+    }
+
+    @Override
+    public void componentClosed() {
+        dataObject.getPrimaryFile().removeFileChangeListener(fcl);
+    }
+
+    @Override
+    public void componentShowing() {
+    }
+
+    @Override
+    public void componentHidden() {
+    }
+
+    @Override
+    public void componentActivated() {
+    }
+
+    @Override
+    public void componentDeactivated() {
+    }
+
+    @Override
+    public UndoRedo getUndoRedo() {
+        return UndoRedo.NONE;
+    }
+
+    @Override
+    public void setMultiViewCallback(MultiViewElementCallback callback) {
+    }
+
+    @Override
+    public CloseOperationState canCloseElement() {
+        return CloseOperationState.STATE_OK;
+    }
+
+    private void addMouseWheelListenerToViewer(JScrollPane scrollPane) {
+        svgPanel.addMouseWheelListener(e -> {
+            double oldScale = scale;
+
+            // Point in scrolled pane
+            Point visiblePoint = e.getPoint();
+
+            // "Picturepixel"
+            Point markedPoint = new Point(
+                (int) (visiblePoint.getX() / oldScale),
+                (int) (visiblePoint.getY() / oldScale));
+
+            int clicks = e.getWheelRotation();
+            int clicks_abs = Math.abs(clicks);
+            for (int i = 0; i < clicks_abs; i++) {
+                if (clicks < 0) {
+                    zoomIn();
+                } else {
+                    zoomOut();
+                }
+            }
+
+            double newScale = scale;
+
+            Point markedPointInRealSpace = new Point(
+                (int) (markedPoint.getX() * newScale),
+                (int) (markedPoint.getY() * newScale)
+            );
+
+            Rectangle r = scrollPane.getViewport().getViewRect();
+
+            r.setLocation(markedPointInRealSpace);
+            r.translate(-r.width / 2, -r.height / 2);
+
+            svgPanel.scrollRectToVisible(r);
+        });
+    }
+
+    private void updateView() {
+        FileObject fo = dataObject.getPrimaryFile();
+
+        if (fo == null) {
+            return;
+        }
+
+        if (viewer == null) {
+            return;
+        }
+
+        svgDocument = svgLoader.load(fo.toURL());
```

Comment:
not sure what the failure mode of SVGLoader is. Does it return null? Throw runtime exceptions? Javadoc unfortunately doesn't mention it. Probably something we should test by trying to load a broken svg file.

It would be better to extract loading into a utility method, which handles logging and Exceptions, so that both `SvgNavigatorPanel` and `SVGViewerElement` can use it.

Reply:
@weisJ thanks for the info! I haven't looked through it using an IDE until now so I wasn't sure about the details.